### PR TITLE
feat: Gate Q combined vol+seasonality continuous IOHMM (D=2)

### LINKED ISSUE #52: Gate Q: Combined vol-ratio + seasonality continuous IOHMM (D=2)
## Gate Q — Combined volatility-ratio + seasonality IOHMM (faithful replication)

Paper §4 evaluates IOHMM-Predictor-I (vol-ratio), IOHMM-Predictor-II (seasonality), **and a combined IOHMM** that conditions transitions on both predictors jointly. The repo currently has both single-predictor variants in bucketed and HMC-continuous form (Gate H + Gate K, PR #48), but the **combined variant is missing** — and is explicitly flagged in both [`docs/paper_pipeline_walkthrough.md`](../blob/main/docs/paper_pipeline_walkthrough.md) and [`docs/results_vs_paper.md`](../blob/main/docs/results_vs_paper.md) as the single most likely change to lift pre-cost Sharpe on this dataset.

This issue closes that §4 coverage gap by **generalising the existing Gate K continuous-parametric form from scalar `x_t ∈ ℝ` to vector `x_t ∈ ℝ²`** carrying vol-ratio and seasonality jointly.

## Why combined, and why now

- **Closes a paper §4 coverage gap, not a refinement.** The trading-model table in the paper has the combined variant; the repo doesn't. After Gate N landed (#51), the headline table is 5/6 paper models on the trading side; Gate Q is one of the remaining coverage items (along with Gate O's MCMC HMM).
- **Highest-leverage Sharpe candidate documented on this data.** Vol-ratio and seasonality individually move pre-cost Sharpe from 0.5298 (baseline) to 0.7577 and 0.6285 respectively. Every ablation tried so far — conviction weighting, K sweep, h=60 — failed. None of them tested **joint conditioning**. Cross-effects ("high vol-ratio behaves differently at the open than at midday") are the unexplored axis.
- **Methodologically distinct from Gate K.** New transition structure (vector side info), not a new inference method. NUTS continues to sample only `(W, b)`; §2.5 exclusions (no MCMC on Θ, no bridge sampling, no asynchronous IOHMM) all stand.

## Model

The Gate K model generalises naturally:

$$P(m_t = n \mid m_{t-1} = m, x_t) = \mathrm{softmax}_n\bigl(W_m \cdot x_t + b_m\bigr), \qquad x_t \in \mathbb{R}^D$$

where `W ∈ ℝ^{K × K × D}` and `b ∈ ℝ^{K × K}`. For `D = 2`, `K = 2` that is 4·2 + 4 = 12 free softmax parameters before row-centering, ~9 after — still trivially within NUTS reach.

The existing Gate K module (`src/hft_hmm/models/iohmm_continuous.py`) implements the scalar case. This issue **generalises that module to support arbitrary D**, with existing D=1 callers (the two HMC single-predictor variants) producing **bit-for-bit identical** posterior samples and diagnostics.

## Acceptance criteria (from `IMPLEMENTATION_PLAN.md` Gate Q)

- [ ] A new variant `vol_ratio_seasonality_hmc_continuous` is registered in `EXPECTED_VARIANTS`, fit via NumPyro NUTS with `A_ij(x_t) = softmax_j(W_i · x_t + b_i)` where `x_t ∈ ℝ²` carries vol-ratio and seasonality jointly.
- [ ] The `iohmm_continuous` module is generalised from scalar `x_t` to vector `x_t` so the same forward filter, posterior-mean transition function, and convergence diagnostics are reused — **no parallel code path**. Existing D=1 callers remain bit-for-bit identical (verified by a regression test).
- [ ] Per-feature standardization on the training slice (each component of `x_t` has its own training mean/std). Leakage-free.
- [ ] Convergence is gated by the existing `ContinuousIOHMMConfig` thresholds; per-window divergent fits flagged as in Gate K.
- [ ] Tests verify (a) D=1 collapses to the existing Gate K result on the single-feature variants, (b) combined transitions are row-stochastic and finite, (c) standardization fits only on the training slice, (d) the variant uses the existing forward filter and walk-forward rig with no parallel code path.

## Evidence expected in PR review

- [ ] A new row for `vol_ratio_seasonality_hmc_continuous` in the headline table or in a dedicated combined-IOHMM subsection of `docs/results_vs_paper.md`.
- [ ] Side-by-side comparison: combined vs vol-ratio HMC vs seasonality HMC pre-cost Sharpe, with an interpretation of whether joint conditioning captures information neither single predictor catches alone.
- [ ] Per-window convergence summary (R-hat, ESS) for the combined variant, matching the Gate K reporting convention.
- [ ] A reproducibility check via `scripts/repro.py`.

### Tasks

- Generalise `src/hft_hmm/models/iohmm_continuous.py` from scalar to vector side information:
  - Accept `side_information` of shape `(T,)` (D=1, legacy) **or** `(T, D)`.
  - `W` parameter shape becomes `(K, K, D)`; `b` stays `(K, K)`. Standardization stored as two `(D,)` arrays.
  - `transition_probabilities_at` accepts `(N,)` (D=1) or `(N, D)` `x_values`.
  - `fit_continuous_iohmm` and `ContinuousIOHMMResult` keep the same public surface; only shapes generalise.
- Register `vol_ratio_seasonality_hmc_continuous` in `EXPECTED_VARIANTS` and extend `_HMC_CONTINUOUS_VARIANTS` / `_SIDE_INFO_VARIANTS`.
- Extend `_build_feature` to return a 2-column `pd.DataFrame` for the combined variant; thread that through the side-info comparison runner so the existing `_dynamic_forward_expected_returns_continuous` helper consumes vector features without a parallel code path.
- Add a config under `configs/example_es_databento_side_info_comparison_combined.yaml` mirroring the existing HMC config layout.
- Update `docs/results_vs_paper.md` and `docs/paper_pipeline_walkthrough.md` to reflect the variant landing (remove the "deferred" framing for Gate Q).

### Tests

- New file or additions in `tests/test_iohmm_continuous.py`:
  - **D=1 backward compat:** posterior samples on a fixed-seed D=1 fit are bit-for-bit identical before and after the refactor (regression test).
  - **D=2 row-stochasticity + finiteness:** evaluate `transition_probabilities_at` on a grid of `(vol_ratio, seasonality)` pairs and assert each row sums to 1 and contains only finite entries.
  - **Synthetic recovery (D=2):** simulate from a known `(W, b)` with `D=2`, fit, assert posterior means recover the true parameters within 2σ.
  - **Per-feature standardization:** `standardization` exposes one `(mean, std)` per feature; both are computed from the training slice only.
- New integration test in `tests/test_side_info_comparison.py` (or equivalent): the combined variant runs end-to-end on the existing fixture, produces a `ContinuousSideInfoVariantWindow`, and writes the expected artifact fields.

### Deliverables

- updated `src/hft_hmm/models/iohmm_continuous.py` (generalised, D=1 callers unchanged)
- updated `src/hft_hmm/experiments/side_info_comparison.py` registering the new variant and routing vector features through the existing continuous helper
- new tests in `tests/test_iohmm_continuous.py` and `tests/test_side_info_comparison.py`
- new config `configs/example_es_databento_side_info_comparison_combined.yaml`
- one tracked `runs/<run_id>/` produced via `scripts/repro.py` on the new config (headline rerun is separate — see "Cost" below)
- a `docs/results_vs_paper.md` row + paragraph for the new variant
- updated `docs/paper_pipeline_walkthrough.md` reflecting Gate Q landing

## Out of scope

- HMC over the Gaussian HMM emission parameters Θ — remains excluded by §2.5 (tracked as Gate O).
- Bridge sampling for K selection — excluded by §2.5 (tracked as Gate P).
- Asynchronous IOHMM, multi-asset, production execution — excluded by §2.5.
- A *bucketed* combined variant. Bucketed-A on `x_t ∈ ℝ²` would require a 2-D bucket grid that explodes the per-bucket sample count; the continuous form is the natural path for D > 1 and is what the paper specifies anyway.

## Resolved decisions

These mirror the Gate K resolved decisions (issue #47) and remain the contract for Gate Q:

1. **Emissions fixed at the per-window Baum-Welch fit.** NUTS samples only `(W, b)`. No co-sampling of Θ.
2. **Per-window retrain.** One HMC fit per walk-forward window on the training slice. At K ∈ {2, 3} with D=2, each fit is still seconds.
3. **Weakly informative priors.** `W_{ijd} ~ Normal(0, 1)`, `b_{ij} ~ Normal(0, 1)` on per-feature standardised `x_t`. Same defensibility argument as Gate K.
4. **Row-centring convention.** Same post-hoc row-mean removal as Gate K — `W` and `b` returned with zero-sum-per-row over the *to-state* axis. R-hat / ESS computed on centred samples.

## Branch

Suggested: `feat/52-combined-iohmm`.

## Cost

~1 week of implementation + tests + docs, plus one ~10-20 h headline rerun for the combined-IOHMM artifact (run separately from this PR; this PR ships the *capability*).

## References

- `IMPLEMENTATION_PLAN.md` Gate Q (lines 469–502)
- `docs/paper_pipeline_walkthrough.md` §"Side-information predictors" (combined variant flagged)
- `docs/results_vs_paper.md` "Sharpe-improvement experiments" (combined IOHMM as highest-leverage candidate)
- Issue #47 / PR #48 — Gate K HMC IOHMM scalar form (the module being generalised)
- `src/hft_hmm/models/iohmm_continuous.py` — existing D=1 implementation
- `src/hft_hmm/experiments/side_info_comparison.py` — variant registry + per-window runner

### PR DESCRIPTION
**Pull request by SantiBarrios2002**
Closes #52.

## Summary
- Generalises `iohmm_continuous` from scalar side information to `(T, D)` side information while preserving the original D=1 NumPyro sample-site shape and scalar likelihood path for bit-for-bit Gate K compatibility.
- Adds `vol_ratio_seasonality_hmc_continuous` to the side-info comparison runner, with a two-column `["vol_ratio", "seasonality"]` feature frame and the existing continuous forward filter / `transition_probabilities_at` path.
- Adds the full Databento combined config plus a tiny fixture smoke config. The smoke-run artifacts under `runs/3a83f3a071bc/` are intentionally **not** committed (untracked in 4b22585); the config is marked deliberately non-production (loose convergence gating for a fast end-to-end pipeline check), and production diagnostics come from the full Databento run `22bbd2c8d0a4`.
- Updates Gate Q docs from deferred/scoped to landed capability, **now including the completed 6-year Databento headline rerun (`22bbd2c8d0a4`)**.

## Results — headline rerun `22bbd2c8d0a4` (negative)
The full 6-year Databento ES rerun completed (92 walk-forward windows, K=2, h_days=23, sign policy). The combined variant is a **negative result**: joint conditioning on `x_t ∈ ℝ²` underperforms *both* single predictors and sits barely above the plain baseline.

| Variant | Pre-cost Sharpe |
|---|---:|
| Volatility-ratio IOHMM (bucketed) | **0.7583** |
| Volatility-ratio IOHMM (HMC) | 0.6515 |
| Seasonality IOHMM (bucketed) | 0.6287 |
| Seasonality IOHMM (HMC) | 0.6281 |
| **Combined vol+seasonality IOHMM (HMC)** | **0.5453** |
| Baseline HMM | 0.5298 |
| Default HMM (weak floor) | −0.1132 |

- Combining the predictors **diluted** the vol-ratio edge rather than stacking the two signals; the predicted cross-effect lift did not materialise. The best configuration found remains **vol-ratio bucketed** (0.7583).
- All 92 windows converged cleanly (rhat ≈ 1.00–1.003, ess healthy), so the result is not a sampler artifact — the doubled input dimension fits fine, it just doesn't help the trading signal.
- Single-predictor variants reproduce the Gate K run `d8b6e7eef6c2` within numerical noise (vol-ratio bucketed 0.7583 vs 0.7577; vol-ratio HMC 0.6515 vs 0.6513).
- `docs/results_vs_paper.md` is updated with the row + run entry, and the now-falsified "combined is the single most likely Sharpe lift" prediction in the next-levers section is corrected. This is the project's third consistent negative result (with Gate K continuous-vs-bucketed and conviction-weighting), reinforcing that the paper-faithful `sign` / `K=2` / `h_days=23` config is a robust local optimum on this data.

## Backward-Compat Evidence
- Added a fixed-seed D=1 regression test that compares stored pre-refactor posterior `W`, `b`, and diagnostics bit-for-bit after squeezing the new singleton feature axis.
- The D=1 NumPyro model keeps the original `W` sample-site shape and scalar `lax.scan` likelihood expression; the `(K, K, 1)` public shape is added only after samples are returned.

## Tests
- ` .venv/bin/ruff check .`
- `find src tests scripts -name '*.py' -exec .venv/bin/python -m black --check {} \;`
- `timeout 180 .venv/bin/black --check .` printed `64 files would be left unchanged` but did not exit before timeout in this sandbox.
- `.venv/bin/pytest tests/test_iohmm_continuous.py tests/test_side_info_comparison.py -x`
- `.venv/bin/python scripts/repro.py configs/gate_q_combined_smoke.yaml --force`

## Summary by CodeRabbit

* **New Features**
  * Added a combined volatility-ratio + seasonality HMC variant (Gate Q) with example run configurations and multi-feature side-information support.

* **Documentation**
  * Walkthroughs, results, and implementation plan updated to mark the combined variant as landed and report its 6‑year rerun (negative headline outcome) and convergence notes.

* **Tests**
  * Expanded coverage for 1D/2D multi-feature continuous models, persistence/shape expectations, and backward-compatibility checks.

### CODE REVIEW

**Comment by SantiBarrios2002** [comment]
## Gate Q headline results landed (run `22bbd2c8d0a4`)

The full 6-year Databento ES rerun completed. Combined vol-ratio + seasonality HMC continuous IOHMM is a **negative result**:

| Variant | Pre-cost Sharpe |
|---|---:|
| Volatility-ratio IOHMM (bucketed) | **0.7583** |
| Volatility-ratio IOHMM (HMC) | 0.6515 |
| Seasonality IOHMM (bucketed) | 0.6287 |
| Seasonality IOHMM (HMC) | 0.6281 |
| **Combined vol+seasonality IOHMM (HMC)** | **0.5453** |
| Baseline HMM | 0.5298 |

Joint conditioning on `x_t ∈ ℝ²` (0.5453) underperforms *both* single predictors and sits barely above baseline — it diluted the vol-ratio edge rather than stacking the signals. All 92 windows converged cleanly (rhat ≈ 1.00–1.003), so this is a real finding, not a sampler artifact. `docs/results_vs_paper.md` is updated with the row, the run entry, and a correction to the now-falsified "combined is the most likely Sharpe lift" prediction.

File: runs/3a83f3a071bc/seasonality_hmc_continuous.log.jsonl [hunk 1]
```diff
@@ -0,0 +1,2 @@
...
+{"chosen_k": 2, "converged": false, "ess_bulk_min": 0, "forecast_end": "2024-01-29T21:15:00+00:00", "forecast_start": "2024-01-23T07:00:00+00:00", "index": 1, "n_forecast_obs": 4280, "n_train_obs": 8364, "posterior_mean_W": [[[0.7775987386703491], [-0.7775987386703491]], [[0.7719261646270752], [-0.7719261646270752]]], "posterior_mean_b": [[-0.0381392240524292, 0.0381392240524292], [0.6063820123672485, -0.6063820123672485]], "rhat_max": 10.0, "summary": {"post-cost": {"cost_bps_per_turnover": 1.0, "cumulative_return": 0.01412921, "hit_rate": 0.44286048, "max_drawdown": -0.00866657, "n_periods": 4279.0, "sharpe_ratio": 0.0171144}, "pre-cost": {"cost_bps_per_turnover": 0.0, "cumulative_return": 0.01412921, "hit_rate": 0.44286048, "max_drawdown": -0.00866657, "n_periods": 4279.0, "sharpe_ratio": 0.0171144}}, "train_end": "2024-01-22T21:15:00+00:00", "train_start": "2024-01-09T07:00:00+00:00"}
```
[the remark this answers is not in the document]
Comment (SantiBarrios2002) [reply runs/3a83f3a071bc/seasonality_hmc_continuous.log.jsonl]: Thanks — this was already addressed in an earlier review round and the artifacts are no longer part of the PR:

- **Smoke outputs are no longer committed.** `runs/3a83f3a071bc/` was untracked in 4b22585 ("Untrack the underpowered Gate Q smoke-run artifacts"); `git ls-tree HEAD runs/3a83f3a071bc/` is empty. This comment is on a path that isn't in the PR anymore.
- **Already marked non-production in metadata.** The smoke config's `notes` field states the convergence gating is deliberately loose (3 warmup / 4 samples, rhat_threshold 10, ess_bulk_threshold 1) and that diagnostics are non-production / outputs not committed.
- **Raising NUTS iterations is intentionally out of scope here.** This config only verifies the pipeline runs end-to-end fast in CI. Production-quality diagnostics come from the full Databento run `22bbd2c8d0a4`: all 92 windows converged (rhat ≈ 1.00–1.003, ess in the thousands).
- **No per-window cache to guard.** Per-stage checkpointing only activates with `--checkpoint`; the smoke path (`repro.py … --force`) refits every window on every run.

Resolving as already-addressed.
